### ISSUE TITLE
Plugin acquisition: Do not install older versions if a newer version is determined

### ISSUE THREAD

**Issue by justinvp** [pulumi/pulumi#12177]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

The engine attempts to determine the required plugins by 1) looking at plugins in the state and (via `gatherPluginsFromSnapshot`) and 2) asking the language host what plugins are required by the program (via `gatherPluginsFromProgram`). This can lead to having two different versions of a plugin for a provider downloaded, an older version specified in state, and a new version specified by the program, when really the expectation is that only the version specified by the program should be used.

To address this, we'll update the default plugin acquisition logic so that the only time we should use the old provider version is for deletes iff we haven't been given a new provider version to use.

**Comment by AaronFriel** [comment]
Would this enable provider maintainers to retract bad releases by deleting or blocking downloads of the plugin, and asking uses to update to a newer SDK?

Context: the ecosystem team had to yank releases of TLS and Random to prevent bad replaces, resulting in user reports such as pulumi/pulumi-random#270 and pulumi/pulumi-random#265. 

----

Related, but further out: We have a design doc forthcoming for how to make plugin acquisition smarter in these cases, allowing providers to publish plugins in a manner that lets us mark a plugin as "retracted" in a way that the host could act on - e.g.: by suggesting the user use a newer version, or selecting an automatic replacement.

**Comment by Frassle** [comment]
> Would this enable provider maintainers to retract bad releases by deleting or blocking downloads of the plugin, and asking uses to update to a newer SDK?

No, this is about resolving plugins when looking at state and the programs GetDependencies output. Currently if state says load v1.0 and GetDependencies says v2.0 we'll load both. This is to change it just load v2.0.

**Comment by AaronFriel** [comment]
Right, if the user updates their SDK to v2.0, they won't receive errors like in the linked issues?

**Comment by Frassle** [comment]
Yes this will help in that situation. It won't do anything for the user who has already downloaded the 1.0 plugin via the 1.0 SDK, which is then yanked and they stay on 1.0 SDK.

**Comment by AaronFriel** [comment]
Got it, it's really helpful though that it could give us a workaround - `pulumi plugin rm ...` and go.

I'm unsure if plugin selection should be context-dependent, but the current behavior is difficult to work around without state edits.

**Comment by lblackstone** [comment]
I wanted to drop a link to https://github.com/pulumi/pulumi/pull/12196 in case we want to pick this change up as well. If not, it can be closed.

**Comment by justinvp** [comment]
I don't think the original issue was fully addressed by #12657.

Here's a repro:

```
mkdir example && cd example && pulumi new random-typescript -g
```

Modify `package.json` to specify an explicit version for `@pulumi/random`, e.g.:

```json
{
  "name": "plug",
  "main": "index.ts",
  "devDependencies": {
    "@types/node": "^14"
  },
  "dependencies": {
    "@pulumi/pulumi": "^3.19.0",
    "@pulumi/random": "4.5.0"
  }
}
```

```
npm install
pulumi up --skip-preview
```

Modify `package.json` again, this time specifying a newer version of `@pulumi/random`:

```json
{
  "name": "plug",
  "main": "index.ts",
  "devDependencies": {
    "@types/node": "^14"
  },
  "dependencies": {
    "@pulumi/pulumi": "^3.19.0",
    "@pulumi/random": "4.13.2"
  }
}
```

```
npm install
pulumi plugin rm resource random --yes
pulumi up
```

It installs both the old version (that it found from the snapshot) and the new version that it's getting from the language host's `GetRequiredPlugins`:

```
Downloading plugin: 13.35 MiB / 13.35 MiB [=========================] 100.00% 6s
                                                                                [resource plugin random-4.5.0] installing
Downloading plugin: 15.00 MiB / 15.00 MiB [=========================] 100.00% 6s
                                                                                [resource plugin random-4.13.2] installing
```

I would expect it to only install the newer version (v4.13.2).

**Comment by Frassle** [comment]
We don't know at setup time (when we do those installs) if the program really only uses 4.13.2 or if its been written to use 4.5.0 and 4.13.2. 
So our choices are three:
1) Greedily try install everything that it looks like we _might_ need, but some might not be used (N.B. if any of these installs fail we don't trigger a stop error)
2) Don't install anything upfront. Let the installer kick in during program execution as and when plugins are actually needed.
3) Try to install the highest version of every plugin we might need, in the hopes that the older versions won't be needed. Sort of half way between option 1 and 2.

In the PR to change this I had left things with option 1, i.e. no change to install behavior, it just changed if those plugin actually tried to load or not. Option 2 doesn't seem great, pushing a lot of fragile download (not because the code here is bad, just because its the internet) to always run during updates. Option 3 seemed prone to cause confusion for the times when old plugins are needed (and this is quite often due to deletes/replaces etc).

So we can change to 3 if it's felt that's the right option, but I think the current state of option 1 is fine.

**Comment by Frassle** [comment]
@justinvp can we get someone to look decided on my comment above. I feel this issue should be closed with the current behavior as is, but if the customer isn't happy with that behavior we need an agreement on what the right behavior _should_ be here.

I'll add that I think the only options are the three I listed above, or a forth which is to stop using the provider listed in state for deletes and instead use the most recent provider. I think that's pretty much a non-starter given worries around providers removing resource types but it is a choice.
